python/pyarrow/types.pxi

```diff
@@ -522,14 +563,29 @@ cdef class StructType(DataType):
         for i in range(len(self)):
             yield self[i]
 
+    def _field(self, int i):
+        """
+        Select a field by its numeric index.
+
+        Parameters
+        ----------
+        i : int
+
+        Returns
+        -------
+        pyarrow.Field
+        """
+        cdef int index = <int> _normalize_index(i, self.struct_type.num_fields())
+        return pyarrow_wrap_field(self.struct_type.field(index))
+
     def __getitem__(self, i):
         """
         Return the struct field with the given index or name.
         """
         if isinstance(i, (bytes, str)):
             return self.field_by_name(i)
         elif isinstance(i, int):
-            return self.field(i)
+            return self._field(i)
```

Comment:
IIUC, the reason for definition of `_field()` is that our new `field()` method masked the parent class implementation, right? I wonder if this works just as well?

```suggestion
            return super().field(i)
```